sql/src/main/java/io/crate/execution/jobs/PageBucketReceiver.java

```diff
@@ -24,53 +24,316 @@
+                pageResultListener.needMore(false);
+            }
+        }
+        boolean shouldTriggerConsumer = false;
+        synchronized (lock) {
+            traceLog("method=setBucket", bucketIdx);
+
+            if (bucketsByIdx.putIfAbsent(bucketIdx, rows) != null) {
+                processingFuture.completeExceptionally(new IllegalStateException(String.format(Locale.ENGLISH,
+                    "Same bucket of a page set more than once. node=%s method=setBucket phaseId=%d bucket=%d",
+                    nodeName, phaseId, bucketIdx)));
+            }
+            setExhaustedUpstreams();
+            if (isLast) {
+                exhausted.add(bucketIdx);
+            }
+            if (bucketsByIdx.size() == numBuckets) {
+                shouldTriggerConsumer = true;
+            }
+        }
+        if (shouldTriggerConsumer) {
+            mergeAndTriggerConsumer();
+        } else if (isLast) {
+            // release listener early here, otherwise other upstreams will be blocked
+            // e.g. if 2 downstream contexts are used in the chain
+            //      Phase -> DistributingDownstream -> Phase -> DistributingDownstream
+            pageResultListener.needMore(false);
+        }
+    }
+
+    private void triggerConsumer() {
+        boolean invokeConsumer = false;
+        Throwable throwable;
+        // TODO: do we really need to synchronise here?
+        synchronized (lock) {
+            if (receivingFirstPage) {
+                receivingFirstPage = false;
+                invokeConsumer = true;
+            }
+            throwable = lastThrowable;
+        }
+        final Throwable error = throwable;
+        if (invokeConsumer) {
+            try {
+                executor.execute(this::consumeRows);
+            } catch (EsRejectedExecutionException | RejectedExecutionException e) {
+                consumer.accept(null, e);
+                throwable = e;
+            }
+        } else {
+            try {
+                executor.execute(() -> batchPagingIterator.completeLoad(error));
+            } catch (EsRejectedExecutionException | RejectedExecutionException e) {
+                batchPagingIterator.completeLoad(e);
+                throwable = e;
+            }
+        }
+        if (throwable != null) {
+            processingFuture.completeExceptionally(throwable);
+        }
+    }
+
+    private void mergeAndTriggerConsumer() {
+        try {
+            mergeBuckets();
+        } catch (Throwable t) {
+            kill(t);
+            // the iterator already returned it's loadNextBatch future, we must complete it exceptionally
+            batchPagingIterator.completeLoad(t);
+            return;
+        }
+        if (allUpstreamsExhausted()) {
+            pagingIterator.finish();
+        }
+        triggerConsumer();
+    }
+
+    private void mergeBuckets() {
+        List<KeyIterable<Integer, Row>> buckets = new ArrayList<>(numBuckets);
+        synchronized (lock) {
```

Comment:
i think this should be locked on `bucketsByIdx` instead as `lock` is already used for a different purpose

Reply:
hmm, we never sync on `bucketByIdx` so far, so I don't think that's the case (`lock` protects `bucketsByIdx` everywhere as far as I can see)

Reply:
yes right it protects it also. could be changed imho, but would leave it now as it is.